torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,111 @@
+import math
+
+import torch
+import torch.distributed as dist
+
+
+def _orthogonalize(matrix, epsilon=1e-8):
+    """Applies Gram-Schmidt procedure to orthogonalize a given 2D tensor."""
+    num_cols = matrix.shape[1]
+    for i in range(num_cols):
```

Comment:
This appears like a procedure that's only dependent on the ith col of the matrix, shouldn't we be able to use some sort of parallelism here (`at::parallel_for` comes to mind, but I'm not an expert on PT threading and parallelism) to speed this procedure up if it were implemented in C++ ?

Reply:
Actually there is only 1 column by default, and at most 2 columns at most for good compression performance.

Reply:
The for-loop is unfortunately inherently sequential here. Each iteration changes the data that will be used by the next one.

Reply:
If the compression performance (speed) hinges on this orthogonalization step, it may be a good idea to try a C++ implementation as @rohan-varma suggested above. Not sure if rank 2 would be high enough *in general* to train a model to full accuracy in the same number of steps as with full-precision communication.

Reply:
If the rank is high, the column-wise scanning can hit a performance issue even implemented in C++, as it won't be cache friendly. Then I think we can further investigate your follow-up work that runs rank-1 compression multiple times.

For now I want to see the max speedup of PowerSGD by using the most aggressive compression setup.